File: py-packages/qgis-sdk/src/qgis_sdk/bridge/window.py
"""qgis_sdk.bridge.window — Window-like EventTarget for Python side.

Provides EventSource/WebSocket-like API for Python to dispatch events to JS.

Python side:
    window = bridge_runtime.to_window()
    window.add_event_listener("message", lambda e: print(e.data))
    window.send({"status": "ready"})
    window.dispatch_event("layer_changed", {"layer_id": "test"})
    window.post_message({"action": "buffer"})

JS side counterpart is QgisBridge extends EventTarget in @qgis-sdk/bridge.
"""

from __future__ import annotations

import json
from typing import Any, Callable, Dict, List, Optional


class Event:
    def __init__(self, type: str, detail: Any = None, data: Any = None):
        self.type = type
        self.detail = detail
        self.data = data if data is not None else detail


class EventTarget:
    """Minimal EventTarget like browser."""

    def __init__(self):
        self._listeners: Dict[str, List[Callable]] = {}

    def add_event_listener(self, event: str, callback: Callable):
        self._listeners.setdefault(event, []).append(callback)

    def addEventListener(self, event: str, callback: Callable, options: Any = None):
        self.add_event_listener(event, callback)

    def remove_event_listener(self, event: str, callback: Callable):
        if event in self._listeners:
            try:
                self._listeners[event].remove(callback)
            except ValueError:
                pass

    def removeEventListener(self, event: str, callback: Callable):
        self.remove_event_listener(event, callback)

    def dispatch_event(self, event: str, data: Any = None) -> bool:
        evt = Event(event, detail=data, data=data)
        listeners = self._listeners.get(event, [])
        for cb in listeners:
            try:
                cb(evt)
            except Exception as e:
                print(f"[BridgeWindow] listener error for {event}: {e}")
        return True

    def dispatchEvent(self, event: Any) -> bool:
        # If event is string, treat as type
        if isinstance(event, str):
            return self.dispatch_event(event)
        # If event has type attribute
        etype = getattr(event, "type", "message")
        detail = getattr(event, "detail", None) or getattr(event, "data", None)
        return self.dispatch_event(etype, detail)


class BridgeWindow(EventTarget):
    """Window-like object for Python → JS communication.

    Wraps a webview and allows sending events to JS via runJavaScript.

    Like WebSocket + EventSource:
    - send(data): sends message event to JS
    - post_message(data): alias
    - dispatch_event(event, data): dispatches CustomEvent to JS
    - close(): closes
    - readyState: CONNECTING/OPEN/CLOSED
    """

    CONNECTING = 0
    OPEN = 1
    CLOSING = 2
    CLOSED = 3

    def __init__(self, runtime: Any = None, webview: Any = None):
        super().__init__()
        self.runtime = runtime
        self.webview = webview
        self.readyState = self.CONNECTING
        self.url = "qgis://bridge"
        self._closed = False

    def _run_js(self, js: str):
        if self.webview is None and self.runtime and hasattr(self.runtime, "webview"):
            webview = self.runtime.webview
        else:
            webview = self.webview

        if webview is None:
            # No webview — just dispatch locally (for testing)
            return

        try:
            # Try QWebEngineView page().runJavaScript
            if hasattr(webview, "page"):
                page = webview.page()
                if hasattr(page, "runJavaScript"):
                    page.runJavaScript(js)
                    return
            # Fallback: webview.runJavaScript
            if hasattr(webview, "runJavaScript"):
                webview.runJavaScript(js)
        except Exception as e:
            print(f"[BridgeWindow] runJavaScript failed: {e}")

    def send(self, data: Any):
        """Send message to JS — like WebSocket send()."""
        try:
            json_data = json.dumps(data)
        except Exception:
            json_data = json.dumps(str(data))
        js = f"""
        (function() {{
            if (window.qgisBridge) {{
                window.qgisBridge.dispatchEvent(new MessageEvent('message', {{data: {json_data}}}));
                if (window.qgisBridge.onmessage) window.qgisBridge.onmessage(new MessageEvent('message', {{data: {json_data}}}));
            }}
            if (window.qgis) {{
                window.qgis.dispatchEvent(new CustomEvent('message', {{detail: {json_data}}}));
            }}
            window.dispatchEvent(new CustomEvent('qgis-bridge-message', {{detail: {json_data}}}));
        }})();
        """
        self._run_js(js)
        # Also dispatch locally for Python listeners
        self.dispatch_event("message", data)

    def post_message(self, data: Any, target_origin: str = "*"):
        self.send(data)

    def postMessage(self, data: Any, target_origin: str = "*"):
        self.send(data)

    def dispatch_event(self, event: str, data: Any = None) -> bool:
        # Dispatch locally first
        super().dispatch_event(event, data)
        # Then dispatch to JS
        try:
            json_data = json.dumps(data) if data is not None else "null"
        except Exception:
            json_data = json.dumps(str(data)) if data is not None else "null"

        js = f"""
        (function() {{
            var detail = {json_data};
            var ev = new CustomEvent('{event}', {{detail: detail}});
            if (window.qgisBridge) window.qgisBridge.dispatchEvent(ev);
            if (window.qgis) window.qgis.dispatchEvent(ev);
            window.dispatchEvent(ev);
        }})();
        """
        self._run_js(js)
        return True

    def close(self, code: int = 1000, reason: str = ""):
        self.readyState = self.CLOSED
        self._closed = True
        js = f"""
        (function() {{
            if (window.qgisBridge) {{
                window.qgisBridge.readyState = 3;
                window.qgisBridge.dispatchEvent(new CloseEvent('close', {{code: {code}, reason: '{reason}'}}));
                if (window.qgisBridge.onclose) window.qgisBridge.onclose(new CloseEvent('close', {{code: {code}, reason: '{reason}'}}));
            }}
        }})();
        """
        self._run_js(js)
        self.dispatch_event("close", {"code": code, "reason": reason})

    def open(self):
        self.readyState = self.OPEN
        js = """
        (function() {
            if (window.qgisBridge) {
                window.qgisBridge.readyState = 1;
                window.qgisBridge.dispatchEvent(new Event('open'));
                if (window.qgisBridge.onopen) window.qgisBridge.onopen(new Event('open'));
            }
        })();
        """
        self._run_js(js)
        self.dispatch_event("open", {})

    def __repr__(self):
        return f"<BridgeWindow readyState={self.readyState} url={self.url}>"


# For backwards compat
Window = BridgeWindow

File: packages-bun/qgis-sdk-bridge/src/window.ts
/** qgis-sdk-bridge window.ts — QgisBridge extends EventTarget, windows-like (EventSource/WebSocket) */

import type { BridgeDescription } from './description';
import { loadDescription } from './description';

export interface BridgeOptions {
  objectName?: string;
  qwebchannelSources?: string[];
  timeout?: number;
  enableQgisApi?: boolean;
  qgisObjectName?: string;
}

export class QgisBridge extends EventTarget {
  // Like WebSocket readyState
  static CONNECTING = 0;
  static OPEN = 1;
  static CLOSING = 2;
  static CLOSED = 3;

  readyState: number = QgisBridge.CONNECTING;
  url: string = 'qgis://bridge';
  description: BridgeDescription | null = null;

  // Properties like window / WebSocket / EventSource
  onopen: ((ev: Event) => void) | null = null;
  onmessage: ((ev: MessageEvent) => void) | null = null;
  onerror: ((ev: Event) => void) | null = null;
  onclose: ((ev: CloseEvent) => void) | null = null;

  // Dynamic bridge methods
  [method: string]: any;

  private _objectName: string;
  private _options: BridgeOptions;
  private _rawBridge: any = null;
  private _rawQgis: any = null;
  private _channel: any = null;

  constructor(objectName = 'bridge', options: BridgeOptions = {}) {
    super();
    this._objectName = objectName;
    this._options = options;
    this.url = `qgis://${objectName}`;
  }

  get objectName() {
    return this._objectName;
  }

  get rawBridge() {
    return this._rawBridge;
  }

  get rawQgis() {
    return this._rawQgis;
  }

  async _connect(): Promise<void> {
    const { loadQWebChannel } = await import('./loader.js');
    
    // Try to load description from window first (injected by Python)
    this.description = loadDescription();
    
    await loadQWebChannel(this._options.qwebchannelSources);

    return new Promise((resolve, reject) => {
      const timeout = this._options.timeout || 10000;
      let timeoutId: any;

      if (timeout > 0) {
        timeoutId = setTimeout(() => {
          this.readyState = QgisBridge.CLOSED;
          const err = new Error(`QWebChannel bridge '${this._objectName}' timeout after ${timeout}ms`);
          this.dispatchEvent(new CustomEvent('error', { detail: err }));
          if (this.onerror) this.onerror(new Event('error'));
          reject(err);
        }, timeout);
      }

      try {
        const QWebChannelGlobal = (typeof window !== 'undefined' ? (window as any).QWebChannel : null) || (typeof (globalThis as any).QWebChannel !== 'undefined' ? (globalThis as any).QWebChannel : null);
        
        if (!QWebChannelGlobal) {
          clearTimeout(timeoutId);
          const err = new Error('QWebChannel not available after loading qwebchannel.js');
          this.readyState = QgisBridge.CLOSED;
          this.dispatchEvent(new CustomEvent('error', { detail: err }));
          reject(err);
          return;
        }

        const transport = (typeof window !== 'undefined' ? (window as any).qt?.webChannelTransport : null) || (typeof (globalThis as any).qt !== 'undefined' ? (globalThis as any).qt?.webChannelTransport : null);

        if (!transport) {
          if (typeof process !== 'undefined' && (process as any).env?.NODE_ENV === 'test') {
            clearTimeout(timeoutId);
            this.readyState = QgisBridge.OPEN;
            this._setupProxy();
            this.dispatchEvent(new Event('open'));
            if (this.onopen) this.onopen(new Event('open'));
            resolve();
            return;
          }
          // If no transport but description available (e.g., testing without QGIS), still open
          if (this.description) {
            clearTimeout(timeoutId);
            this.readyState = QgisBridge.OPEN;
            this._setupProxy();
            this.dispatchEvent(new Event('open'));
            if (this.onopen) this.onopen(new Event('open'));
            resolve();
            return;
          }
          clearTimeout(timeoutId);
          const err = new Error('qt.webChannelTransport not available — are you running inside QWebEngineView?');
          this.readyState = QgisBridge.CLOSED;
          this.dispatchEvent(new CustomEvent('error', { detail: err }));
          reject(err);
          return;
        }

        new QWebChannelGlobal(transport, (channel: any) => {
          clearTimeout(timeoutId);
          this._channel = channel;
          const raw = channel.objects[this._objectName];
          const rawQgis = channel.objects[this._options.qgisObjectName || 'qgis'];

          if (!raw && !this.description) {
            const err = new Error(`Bridge object '${this._objectName}' not found. Available: ${Object.keys(channel.objects).join(', ')}`);
            this.readyState = QgisBridge.CLOSED;
            this.dispatchEvent(new CustomEvent('error', { detail: err }));
            reject(err);
            return;
          }

          this._rawBridge = raw || {};
          this._rawQgis = rawQgis || null;

          // Load description if not already loaded, or use injected
          if (!this.description) {
            this.description = loadDescription();
          }

          this.readyState = QgisBridge.OPEN;
          this._setupProxy();
          
          this.dispatchEvent(new Event('open'));
          if (this.onopen) this.onopen(new Event('open'));
          
          resolve();
        });
      } catch (e) {
        clearTimeout(timeoutId);
        this.readyState = QgisBridge.CLOSED;
        const err = e as Error;
        this.dispatchEvent(new CustomEvent('error', { detail: err }));
        if (this.onerror) this.onerror(new Event('error'));
        reject(e);
      }
    });
  }

  private _setupProxy() {
    // Create proxy for direct method access: bridge.get_layer() -> call()
    if (!this._rawBridge) return;

    const methodNames = Object.keys(this._rawBridge).filter(k => typeof this._rawBridge[k] === 'function');
    
    methodNames.forEach(methodName => {
      if (this[methodName]) return; // Don't overwrite existing
      this[methodName] = (...args: any[]) => this.call(methodName, ...args);
    });

    // If description available, also add methods from description that may not be in raw yet
    if (this.description) {
      this.description.methods.forEach(m => {
        if (!this[m.name]) {
          this[m.name] = (...args: any[]) => this.call(m.name, ...args);
        }
      });
    }
  }

  async _getRawQgis(): Promise<any> {
    return this._rawQgis;
  }

  // WebSocket-like
  send(data: any): void {
    if (this._rawBridge && typeof this._rawBridge.send === 'function') {
      this._rawBridge.send(data);
      return;
    }
    const ev = new MessageEvent('message', { data });
    this.dispatchEvent(ev);
    if (this.onmessage) this.onmessage(ev);
  }

  close(code?: number, reason?: string): void {
    this.readyState = QgisBridge.CLOSING;
    this.readyState = QgisBridge.CLOSED;
    const ev = new CloseEvent('close', { code: code || 1000, reason: reason || '' });
    this.dispatchEvent(ev);
    if (this.onclose) this.onclose(ev);
  }

  // EventSource-like — already inherits addEventListener/removeEventListener from EventTarget
  // But we provide typed overloads
  addEventListener(type: string, listener: EventListenerOrEventListenerObject | null, options?: boolean | AddEventListenerOptions): void {
    super.addEventListener(type, listener as any, options);
    // Also wire up on* properties
    if (type === 'open' && typeof listener === 'function') {
      // For onopen, we don't auto-set, but we can track
    }
  }

  removeEventListener(type: string, listener: EventListenerOrEventListenerObject | null, options?: boolean | EventListenerOptions): void {
    super.removeEventListener(type, listener as any, options);
  }

  // Window-like
  postMessage(message: any, targetOrigin?: string): void {
    this.send(message);
  }

  // Bridge method call — like fetch but for bridge
  async call<T>(method: string, ...args: any[]): Promise<T> {
    if (!this._rawBridge) {
      // If no raw bridge but in test mode, return mock
      if (this.readyState === QgisBridge.OPEN) {
        console.warn(`[QgisBridge] No raw bridge, returning mock for ${method}`);
        return null as T;
      }
      throw new Error(`Bridge not connected (readyState=${this.readyState})`);
    }

    const rawMethod = this._rawBridge[method];
    if (typeof rawMethod !== 'function') {
      throw new Error(`Method ${method} not found on bridge '${this._objectName}'. Available: ${Object.keys(this._rawBridge).filter(k => typeof this._rawBridge[k] === 'function').join(', ')}`);
    }

    return new Promise<T>((resolve, reject) => {
      try {
        const lastArg = args[args.length - 1];
        const hasCallback = typeof lastArg === 'function';
        if (hasCallback) {
          rawMethod.apply(this._rawBridge, args);
        } else {
          rawMethod.apply(this._rawBridge, [
            ...args,
            (result: any) => {
              try {
                if (typeof result === 'string') {
                  try {
                    const parsed = JSON.parse(result);
                    resolve(parsed as T);
                    return;
                  } catch {
                    // Not JSON
                  }
                }
                resolve(result as T);
              } catch (e) {
                reject(e);
              }
            }
          ]);
        }
      } catch (e) {
        reject(e);
      }
    });
  }

  // Convenience for dispatching from Python
  dispatchEvent(event: Event): boolean {
    const result = super.dispatchEvent(event);
    // Also call on* handlers
    if (event.type === 'open' && this.onopen) {
      this.onopen(event);
    } else if (event.type === 'message' && this.onmessage) {
      this.onmessage(event as MessageEvent);
    } else if (event.type === 'error' && this.onerror) {
      this.onerror(event);
    } else if (event.type === 'close' && this.onclose) {
      this.onclose(event as CloseEvent);
    }
    return result;
  }
}

// Factory — like new EventSource(url) or new WebSocket(url)
export async function createBridge<T = QgisBridge>(
  objectName: string | BridgeOptions = 'bridge',
  options: BridgeOptions = {}
): Promise<T & QgisBridge> {
  const opts: BridgeOptions = typeof objectName === 'string' ? { objectName, ...options } : objectName;
  const name = opts.objectName || (typeof objectName === 'string' ? objectName : 'bridge');
  
  const win = new QgisBridge(name, opts) as T & QgisBridge;
  await win._connect();
  return win;
}
